cypress/e2e/permit2.test.ts

```diff
@@ -0,0 +1,212 @@
+import { MaxUint160, MaxUint256 } from '@uniswap/permit2-sdk'
+
+import { DAI, USDC_MAINNET } from '../../src/constants/tokens'
+
+const APPROVE_BUTTON = '[data-testid="swap-approve-button"]'
+
+// The same tokens & swap-amount combination is used for all permit2 tests
+const INPUT_TOKEN = DAI
+const OUTPUT_TOKEN = USDC_MAINNET
+const TEST_BALANCE_INCREMENT = 0.01
+
+const setupSwap = () => {
+  // Use input/output in search params for faster test setup
+  cy.visit(`/swap/?inputCurrency=${INPUT_TOKEN.address}&outputCurrency=${OUTPUT_TOKEN.address}`, {
+    ethereum: 'hardhat',
+  })
+  cy.get('#swap-currency-input .token-amount-input').clear().type(TEST_BALANCE_INCREMENT.toString())
+}
+
+// Shorthand util for initiating a swap and confirming it
+const swap = () => {
+  // Ensures that enabled after approval before clicking
+  cy.get('#swap-button').should('not.have.attr', 'disabled')
+
+  // Completes the swap
+  cy.get('#swap-button').click()
+  cy.get('#confirm-swap-or-send').click()
+  cy.get('[data-testid="dismiss-tx-confirmation"]').click()
+
+  // There should be a successful Swap notification
+  cy.contains('Swapped').should('exist')
+}
+
+// TODO: Update tests to differentiate between permit2 vs token approval button text once UI is updated to indicate approval step
+describe('Permit2', () => {
+  let owner: string
+  before(() => cy.hardhat().then((hardhat) => (owner = hardhat.wallet.address)))
+
+  beforeEach(setupSwap)
+
+  // Shorthand util for approving the input token, permit2 spend, or both
+  const setApproval = (type: 'token' | 'permit2' | 'both') => {
+    return cy.hardhat().then(async (hardhat) => {
+      const promises: Array<Promise<void>> = []
+      if (type === 'token' || type === 'both') {
+        promises.push(hardhat.approval.setTokenAllowanceForPermit2({ owner, token: INPUT_TOKEN }))
+      }
+      if (type === 'permit2' || type === 'both') {
+        promises.push(hardhat.approval.setPermit2Allowance({ owner, token: INPUT_TOKEN }))
+      }
+
+      await Promise.all(promises)
+      return hardhat
+    })
+  }
```

Comment:
Don't use a shorthand util **with branching logic** in tests. You should inline this in the call sites.